Migrate CollectionRightsBackendApiService to Angular 8

h-ARTS opened the pull request:
## Explanation

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [X] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [X] The PR is made from a branch that's **not** called "develop".
- [X] The PR has an appropriate "PR CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [X] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

h-ARTS committed:
Merge CollectionRightsBackendApiService to Angular 8

* Upgrade service and test file to Angular 8
* $rootScope.$apply() added in the directive where method gets called
* Request Interceptor added to the module
* Test performed with success

h-ARTS committed:
yarn.lock

h-ARTS committed:
max-len fix

nishantwrp commented on yarn.lock:
```diff
@@ -975,19 +975,11 @@ append-transform@^1.0.0:
     default-require-extensions "^2.0.0"
 
-aproba@^1.0.3, aproba@^1.1.1:
+aproba@^1.1.1:
```
This file is not supposed to be changed. Please revert back the file to the original state.

nishantwrp replied:
This file is not supposed to be changed. Please revert back the file to it's original state.

nishantwrp commented on core/templates/domain/collection/collection-rights-backend-api.service.ts:
```diff
@@ -16,143 +16,168 @@
+      collectionVersion: any,
+      isPublic: Boolean,
+      successCallback: any,
+      errorCallback: any): void {
```
Please write the exact types for these arguments.

nishantwrp reviewed (changes requested):
Hi @h-ARTS, Thanks for the PR. Left some comments, PTAL!

h-ARTS committed:
Revert "yarn.lock"

This reverts commit 52bf121fac64b3f7caaf7a85a6d1c788ae08a7f1.

h-ARTS committed:
Forgotten types specified

bansalnitish commented on core/templates/domain/collection/collection-rights-backend-api.service.ts:
```diff
@@ -16,143 +16,168 @@
+   */
+  cacheCollectionRights(collectionId: string,
+      collectionRights: Array<object>): void {
+    this.collectionRightsCache[collectionId] = angular.copy(collectionRights);
```
Please use cloneDeep here.

h-ARTS replied:
Done!

bansalnitish commented on core/templates/domain/collection/collection-rights-backend-api.service.ts:
```diff
@@ -16,143 +16,168 @@
-        });
-      },
+    this.http.put(requestUrl, putParams).toPromise().then((response: any) => {
+      this.collectionRightsCache[collectionId] = response.data;
```
It should be just `response` right, not `response.data`?

Ditto below.

h-ARTS replied:
Yep, I've overseen this as well as with the `error.data`. I'll get this fixed right away 👍

bansalnitish commented on core/templates/domain/collection/collection-rights-backend-api.service.ts:
```diff
@@ -16,143 +16,168 @@
+  'domain/utilities/url-interpolation.service';
+import { CollectionEditorPageConstants } from
+  'pages/collection-editor-page/collection-editor-page.constants';
+import cloneDeep from 'lodash/cloneDeep';
```
Move cloneDeep above all Oppia dependencies. Also, please alphabetise the imports.

h-ARTS replied:
So utilities, in general, should be placed above all other imports?

bansalnitish replied:
```
import { downgrade }..
import { HttpClient } ...

import cloneDeep ...

import oppia dependecies...
```

h-ARTS replied:
Done!

bansalnitish reviewed (commented):
Hi @h-ARTS, this is amazing stuff. It looks like you've followed the docs and previous pull requests nicely :).

I've left some minor nits for you. Please address them. Let me know if you need any help.

h-ARTS committed:
Minor changes

- Imports alphabetically sorted
- cloneDeep used instead of angular.copy
- Response assigned correctly to the collection rights caching (response.data doesn't exist after debugging)
- Error callback returns error object (error.data doesn't exist after debugging)

h-ARTS committed:
imports reordered again according to reviewers comment

DubeySandeep commented:
@h-ARTS, Please merge your PR with upstream/develop to fix the flaky e2e test issue on CircleCI (if it's a flaky error).

h-ARTS commented:
@DubeySandeep I can't merge this PR because I'm not authorized to do that.

kevinlee12 commented:
I think he meant merge from develop.

h-ARTS commented:
> I think he meant merge from develop.

@kevinlee12 But in this case, I have to create a new PR right?

kevinlee12 commented:
no need, just merge from develop and push. Github will automatically update!

bansalnitish commented:
Hey @h-ARTS do let us know if you need any help. Thanks!

h-ARTS commented:
@kevinlee12 @bansalnitish So what I understood so far is to merge from `upstream/develop`(checked out) into `origin/collection-rights-backendapi-service` and then push. If I would do this then I have no permission to push because I'm not authorized:
```
remote: Permission to oppia/oppia.git denied to h-ARTS.
fatal: unable to access 'https://github.com/oppia/oppia.git/': The requested URL returned error: 403
```

PS: gitter seems to be not working at the moment...

seanlip commented:
What do you get when you type `git remote -v` into a terminal?

h-ARTS commented:
@seanlip 
```
origin  git@github.com:h-ARTS/oppia.git (fetch)
origin  git@github.com:h-ARTS/oppia.git (push)
upstream        https://github.com/oppia/oppia.git (fetch)
upstream        https://github.com/oppia/oppia.git (push)
```

seanlip commented:
That seems about right. Are you pushing to origin (h-ARTS/oppia)? You should be doing so, rather than pushing to upstream (oppia/oppia), as stated in your error log.

h-ARTS commented:
@seanlip First I merged from `origin/develop` into `origin/collection-rights-backendapi-service` which of course worked but I can't push `origin/develop` to `upstream/develop` because, sadly, I get the same error:
```
Hanans-MacBook-Pro:oppia hananmufti$ git push upstream develop
remote: Permission to oppia/oppia.git denied to h-ARTS.
fatal: unable to access 'https://github.com/oppia/oppia.git/': The requested URL returned error: 403
```

seanlip commented:
You are not supposed to push to upstream/develop. Please read and follow the [wiki instructions](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change) very carefully.

h-ARTS committed:
Migrate CollectionRightsBackendApiService to Angular 8

* Upgrade service and test file to Angular 8
* $rootScope.$apply() added in the directive where method gets called
* Request Interceptor added to the module
* Test performed with success

h-ARTS commented:
@seanlip Now I pushed again but this time ci/circleci setup and travis is still pending since 2 hours

h-ARTS commented:
@seanlip @nishantwrp @kevinlee12 PTAL

h-ARTS reviewed (commented):
Requested changes are already done

kevinlee12 commented:
@bansalnitish ptal!

h-ARTS committed:
Merge remote-tracking branch 'upstream/develop' into collection-rights-backendapi-service

# Conflicts:
#	core/templates/pages/collection-editor-page/collection-editor-page.module.ts

h-ARTS commented:
@bansalnitish @nishantwrp PTAL!

Please rerun ci/circleci: frontend_tests it's been disconnected.

kevinlee12 commented:
I’ve reran the tests.

> On Mar 9, 2020, at 7:43 AM, Hanan Mufti <notifications@github.com> wrote:
> 
> ﻿
> @bansalnitish @nishantwrp PTAL!
> 
> Please rerun ci/circleci: frontend_tests it's been disconnected.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bansalnitish commented on core/templates/domain/collection/collection-rights-backend-api.service.spec.ts:
```diff
@@ -71,160 +57,200 @@ describe('Collection rights backend API service', function() {
+      flushMicrotasks();
 
-      expect(successHandler).toHaveBeenCalledWith(sampleDataResults);
+      expect(successHandler).toHaveBeenCalled();
```
Please change this to `toHaveBeenCalledWith()` (similar to original code)

bansalnitish commented on core/templates/domain/collection/collection-rights-backend-api.service.spec.ts:
```diff
@@ -16,52 +16,38 @@
+import { HttpClientTestingModule, HttpTestingController } from
+  '@angular/common/http/testing';
+import { TestBed, fakeAsync, flushMicrotasks } from '@angular/core/testing';
+import { CollectionRightsBackendApiService } from
```
Add a line break here.

bansalnitish reviewed (approved):
Some minor nits, LGTM!

h-ARTS committed:
Minor changes

nishantwrp commented on core/templates/domain/collection/collection-rights-backend-api.service.ts:
```diff
@@ -16,143 +16,170 @@
+      collectionVersion: number,
+      isPublic: boolean,
+      successCallback: Function,
+      errorCallback: Function): void {
```
Please write the exact types of successCallback and errorCallback here.

nishantwrp reviewed (changes requested):
Just one comment

seanlip commented:
Hi @bansalnitish @nishantwrp -- please don't forget to update the Assignees list after you comment. Thanks!

h-ARTS committed:
Added exact type of success and error callbacks for setCollectionStatus

h-ARTS commented:
Hi @nishantwrp PTAL
@kevintab95 @aks681

h-ARTS commented:
@kevinlee12 frontend test disconnected again. Please rerun 😥
```
09 03 2020 22:28:59.837:WARN [HeadlessChrome 80.0.3987 (Linux 0.0.0)]: Disconnected (3 times)reconnect failed before timeout of 60000ms (transport close)
```

seanlip commented:
Do the tests run to completion on your local machine?

h-ARTS commented:
@seanlip I did run the frontend test and type checking locally before pushing it to the branch and they were successful.

seanlip commented:
It seems to be successful now. Perhaps it was a CircleCI blip.

kevinlee12 commented:
I'm not sure if this is related to the issue that Vojta was seeing.

On Mon, Mar 9, 2020 at 4:19 PM Sean Lip <notifications@github.com> wrote:

> It seems to be successful now. Perhaps it was a CircleCI blip.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/8748?email_source=notifications&email_token=ABL52MHF2N3TVABULQXSROTRGWBRTA5CNFSM4K7FPUFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOJNFEI#issuecomment-596824721>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABL52MCQX744ES3LNLHHVMDRGWBRTANCNFSM4K7FPUFA>
> .
>

nishantwrp reviewed (approved):
LGTM!

kevintab95 reviewed (approved):
LGTM, thanks @h-ARTS!

h-ARTS commented:
@aks681 PTAL!

bansalnitish commented:
Congrats @h-ARTS for your PR :).
